plugins/plugin-machine/che-plugin-machine-ext-server/src/main/java/org/eclipse/che/ide/ext/machine/server/ssh/KeysInjector.java

```diff
@@ -54,72 +75,100 @@ public void start() {
-                  final List<String> publicMachineKeys = sshPairs.stream()
-                                                       .filter(sshPair -> sshPair.getPublicKey() != null)
-                                                       .map(SshPairImpl::getPublicKey)
-                                                       .collect(Collectors.toList());
+              if (sshWorkspacePair != null && sshWorkspacePair.getPublicKey() != null) {
+                publicKeys.add(sshWorkspacePair.getPublicKey());
+              }
 
-                  // get workspace keypair (if any)
-                  SshPairImpl sshWorkspacePair = null;
-                  try {
-                      sshWorkspacePair = sshManager.getPair(machine.getOwner(), "workspace", event.getWorkspaceId());
-                  } catch (NotFoundException e) {
-                      LOG.debug("No ssh key associated to the workspace", e);
-                  }
+              if (publicKeys.isEmpty()) {
+                return;
+              }
 
-                  // build list of all pairs.
-                  final List<String> publicKeys;
-                  if (sshWorkspacePair != null && sshWorkspacePair.getPublicKey() != null) {
-                      publicKeys = new ArrayList<>(publicMachineKeys.size() + 1);
-                      publicKeys.add(sshWorkspacePair.getPublicKey());
-                      publicKeys.addAll(publicMachineKeys);
-                  } else {
-                      publicKeys = publicMachineKeys;
-                  }
+              StringBuilder commandLine = new StringBuilder("mkdir ~/.ssh/ -p");
+              for (String publicKey : publicKeys) {
+                commandLine
+                    .append("&& echo '")
+                    .append(publicKey)
+                    .append("' >> ~/.ssh/authorized_keys");
+              }
 
-                  if (publicKeys.isEmpty()) {
-                      return;
-                  }
+              ExecAgentClient client = execAgentClientFactory.create(execServer.getUrl());
+              ProcessStartResponseDto startProcess =
+                  client.startProcess(workspaceId, commandLine.toString(), "sshUpload", "custom");
 
-                  final String containerId = machine.getRuntime().getAttributes().get("id");
-                  StringBuilder command = new StringBuilder("mkdir ~/.ssh/ -p");
-                  for (String publicKey : publicKeys) {
-                      command.append("&& echo '")
-                             .append(publicKey)
-                             .append("' >> ~/.ssh/authorized_keys");
-                  }
+              GetProcessResponseDto process = client.getProcess(workspaceId, startProcess.getPid());
+              int tryNumber = 1;
+              while (process.isAlive()) {
+                process = client.getProcess(workspaceId, startProcess.getPid());
+                tryNumber++;
+                if (tryNumber > 10) {
```

Comment:
Are you sure that killing of the process after 10 failed tries is OK? I would say no while there are not any delays between tries.